### ISSUE TITLE
Simplify `ITenantInfo`?

### ISSUE THREAD

**Issue by fdcastel** [Finbuckle/Finbuckle.MultiTenant#836]
1) **Why does `ITenantInfo` have property setters?**

    To the best of my understanding, they are only used in tests.

    `ITenantInfo` interface could be simplified to use only getters. And tests may use a `TestTenantInfo` with setters.

2) While on this page, **why `Identifier` and `Name`?**

    The absolutely minimum required is `Id`. And it should not be nullable.

    Other properties seem to be used only for some specific cases/strategies. It could be refactored.

Sadly, the `7.0` release has already shipped. Maybe in `8.0`? 

I can work on some PRs, if interested.

**Comment by fdcastel** [comment]
Update:

>    To the best of my understanding, they are ~only used in tests~. 

Correction: The _interface_ setters are **not used anywhere**. Tests always use `TenantInfo` class, which _already does_ have setters.



> `ITenantInfo` interface can be simplified to use only getters. 

I did a POC right now in a local fork and **everything builds fine**.
I can provide my own custom `ITenantInfo` implementation without the need for setters.

**Comment by AndrewTriesToCode** [comment]
Yes you are right. I would like to clean this up on the next major release and have ITenantInfo instances be treated more like an immutable.

**Comment by Illivion** [comment]
@AndrewTriesToCode Any movement to ITenantInfo without setters? It should be immutable!

**Comment by fdcastel** [comment]
Hey @AndrewTriesToCode long time no see! 😄

Want me to whip up a PR for this?

**Comment by AndrewTriesToCode** [comment]
Hi, I went to a record based `ITenantInfo` last November and it caused a lot of issues for folks, see #1073. I ultimately reverted it and went with recommendations in the docs to not modify your the tenant info. I’m torn on the issue.

If I had a Time Machine I would have done it this way from the start, but I also don’t want to limit how people use it.

**Comment by fdcastel** [comment]
I fully understand your point.

My original intention with this issue was actually to simplify `ITenantInfo` and reduce the API surface area. However, your suggestion around immutability is also very interesting and, in my opinion, a solid approach as well.

Thanks for the pointers. I’ll take a look at what has been done over the last couple of years to get a better overall perspective.

**Comment by fdcastel** [comment]
Hi @AndrewTriesToCode — as promised, I took some time to go through the history here (#1029 → #1073 → #1075 → #1188 on `next`) before touching anything.

My read, for what it's worth:

- The original asks in this issue (getter-only `ITenantInfo`, no `Name`, non-nullable `Id`) effectively landed in 10.0.2 with #1075, so the interface itself is already where I hoped it would be. Thank you for that.
- What seemed to remain was the tension you described: you'd like the active tenant to be effectively immutable, while #1073 made a good case that the library should only code against the interface and never dictate a concrete shape (record/base type/constructor).

I tried to reconcile those without touching the interface: keep `ITenantInfo` getter-only and shape-agnostic, have the *library* guarantee its side (stores never mutate instances and don't depend on live property reads; the active tenant of a scope is set once — which `next` already does), and back it up with a test matrix that runs every store/cache/resolver/middleware/DbContext against a constructor-only immutable type. Along the way I found that `EchoStore` and `ConfigurationStore` still silently required settable `Id`/`Identifier` via reflection, so a fully immutable implementation couldn't actually be used with them — that felt like the concrete gap behind the "simplify `ITenantInfo`" idea.

I've opened two PRs as a **first attempt**, very much open to discussion:

- #1204 against `next` (v11): contract documented on `ITenantInfo<TId>`, factory overloads + clear errors for `EchoStore`/`ConfigurationStore`, snapshot-keyed `InMemoryStore`, docs, and the immutable-type test matrix.
- #1205 against `main` (10.x): the additive subset only (factory overloads, clear errors, docs, tests), no behavior changes for existing users.

Please treat both as a starting point rather than a finished answer — if you'd prefer a different approach (e.g. only tests + docs, keeping `ConfigurationStore` binding as-is on `next`, or something else entirely), I'm happy to rework or drop parts. And if you feel this issue is already covered by 10.0.2 and the PRs, feel free to close it whenever you like.

Refs #1073, #602.
